openedx/core/djangoapps/waffle_utils/__init__.py

```diff
@@ -223,7 +223,9 @@ def is_flag_active(self, flag_name, check_before_waffle_callback=None, flag_unde
                 if value is None:
                     value = flag_is_active(get_request(), namespaced_flag_name)
 
-            self._cached_flags[namespaced_flag_name] = value
+            # Don't cache the result if it might have been affected by the callback
```

Comment:
Good call @andy-armstrong.  Thanks.  A couple of additional comments on some possible clean-up.

1. I'd change this comment to:
```
# If wanted, caching for the callback must be done by the callback provider, because the
# cache key may require additional name-spacing.  See CourseWaffleFlag.
```
2. Since I wrote that comment, we should add caching to CourseWaffleFlag for fun. :)  The following is untested sample code:
```
cache_key = '{}.{}'.format(self.namespaced_flag_name, course_key.to_deprecated_string())
value = self. waffle_namespace._cached_flags[cache_key]
if value is None:
   ...
   self. waffle_namespace._cached_flags[cache_key] = value
return value
```
3. You missed some renaming of course_id to course_key.
4. Other possible Waffle Flag cleanup is that strictly speaking, the callback doesn't need to send the flag name because WaffleFlag has a namespaced_flag_name property.  If we don't use this property here, should it be deleted if it has no uses?